cts.json

```diff
@@ -2115,6 +2120,226 @@
+      "selector" : "$[?length(1)>=2]",
+      "document" : [{"a": "ab"}, {"a": "d"}],
+      "result": []
+    },
+    {
+      "name": "filter, length function, result must be compared",
+      "selector" : "$[?length(@.a)]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, length function, no params",
+      "selector" : "$[?length()==1]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, length function, too many params",
+      "selector" : "$[?length(@.a,@.b)==1]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, count function",
+      "selector" : "$[?count(@..*)>2]",
+      "document" : [
+        {"a": [1, 2, 3]},
+        {"a": [1], "d": "f"},
+        {"a": 1, "d": "f"}
+      ],
+      "result": [
+        {"a": [1,2,3]},
+        {"a": [1], "d":"f"}
+      ]
+    },
+    {
+      "name": "filter, count function, non-array/string arg",
+      "selector" : "$[?count(1)>2]",
+      "document" : [
+        {"a": [1, 2, 3]},
+        {"a": [1], "d": "f"},
+        {"a": 1, "d": "f"}
+      ],
+      "result": []
+    },
+    {
+      "name": "filter, count function, result must be compared",
+      "selector" : "$[?count(@..*)]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, count function, no params",
+      "selector" : "$[?count()==1]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, count function, too many params",
+      "selector" : "$[?count(@.a,@.b)==1]",
+      "invalid_selector": true
+    },
+    {
+      "name": "filter, match function, found match",
+      "selector" : "$[?match(@.a, 'a.*')]",
+      "document" : [{"a": "ab"}],
+      "result": [
+        {"a":"ab"}
+      ]
+    },
+    {
+      "name": "filter, match function, double quotes",
+      "selector" : "$[?match(@.a, \"a.*\")]",
+      "document" : [{"a": "ab"}],
+      "result": [
+        {"a":"ab"}
+      ]
+    },
+    {
+      "name": "filter, match function, don't select match",
+      "selector" : "$[?!match(@.a, 'a.*')]",
+      "document" : [{"a": "ab"}],
+      "result": [
+      ]
+    },
+    {
+      "name": "filter, match function, not a match",
+      "selector" : "$[?match(@.a, 'a.*')]",
+      "document" : [{"a": "bc"}],
+      "result": []
+    },
+    {
+      "name": "filter, match function, select non-match",
+      "selector" : "$[?!match(@.a, 'a.*')]",
+      "document" : [{"a": "bc"}],
+      "result": [{"a": "bc"}]
+    },
+    {
+      "name": "filter, match function, invalid first arg",
+      "selector" : "$[?match(1, 'a.*')]",
+      "document" : [{"a": "ab"}],
+      "result": []
+    },
+    {
+      "name": "filter, match function, invalid second arg",
+      "selector" : "$[?match(@.a, 1)]",
+      "document" : [{"a": "ab"}],
+      "result": []
+    },
+    {
+      "name": "filter, match function, result cannot be compared",
+      "selector" : "$[?match(@.a, 'a.*')==true]",
+      "invalid_selector": true
```

Comment:
The spec currently allows functions which return `OptionalNodeOrValue` or one of its subtypes to be used in comparisons. `OptionalBoolean` is a subtype of `OptionalNodeOrValue`. Therefore the result of `match` can be compared.

Reply:
`OptionalNodeOrValue` is only defined for functions. Where does the ABNF allow this?

Reply:
Okay...

`function-expr` is the only representation of functions in the ABNF (aside from describing the syntax for functions), and that is permissible in all expressions.

However, that means that there can be no type checking during parsing, so what's the point of having a type system at all?

Reply:
The ABNF doesn't capture the whole story. There is normative text that describes syntax which is valid in the ABNF but invalid in the spec.

Reply:
Yeah, I don't buy this argument.  I've lost so many decisions because what I wanted wasn't expressed in the ABNF.

Reply:
I also argue that `OptionalNodeOrValue` shouldn't include `Boolean`.

Reply:
Any path with a comparison involving a function that returns an `OptionalNodes` has a type error.

Reply:
See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/387

Reply:
Discussion has carried on to https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/404 and https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/405